src/modules/tools/temperaturecontrol/TemperatureControl.cpp

```diff
@@ -347,17 +355,25 @@ void TemperatureControl::on_gcode_received(void *argument)
                         }
 
                         this->waiting = true; // on_second_tick will announce temps
-                        while ( get_temperature() < target_temperature ) {
+                        float t = get_temperature();
+                        while ( t < target_temperature || (r != UNDEFINED && t > r) ) {
```

Comment:
This is not very robust, there are many situations where the user can set values here that will cause it to wait indefinitely. I would add some sanity checks to make sure that not only is r < t but also that it is at least a few degrees less.